contracts/andromeda-portal-ado/src/contract.rs

```diff
@@ -0,0 +1,285 @@
+};
+use cw0::PaymentError;
+use cw2::set_contract_version;
+use cw20::{Cw20Coin, Cw20ReceiveMsg};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "andromeda-portal-ado";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: InitMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    let cfg = Config {
+        default_timeout: msg.default_timeout,
+    };
+    CONFIG.save(deps.storage, &cfg)?;
+
+    for white_addr in msg.whitelist {
+        let contract = deps.api.addr_validate(&white_addr)?;
+        WHITE_LIST.save(deps.storage, &contract, &true)?;
+    }
+
+    // Ok(Response::new().add_attributes(vec![
+    //     attr("action", "instantiate"),
+    //     attr("default_timeout", msg.default_timeout.to_string()),
+    // ]))
+
+    ADOContract::default().instantiate(
+        deps.storage,
+        deps.api,
+        info,
+        BaseInstantiateMsg {
+            ado_type: "portal-ado".to_string(),
+            operators: None,
+            modules: None,
+            primitive_contract: None,
+        },
+    )
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::AndrReceive(msg) => execute_andr_receive(deps, env, info, msg),
+        ExecuteMsg::Receive(msg) => execute_receive(deps, env, info, msg),
+        ExecuteMsg::Transfer(msg) => {
+            let coin = match info.funds.len() {
+                0 => Err(PaymentError::NoFunds {}),
+                1 => {
+                    let coin = &info.funds[0];
+                    if coin.amount.is_zero() {
+                        Err(PaymentError::NoFunds {})
+                    } else {
+                        Ok(coin.clone())
+                    }
+                }
+                _ => Err(PaymentError::MultipleDenoms {}),
+            }?;
+            execute_transfer(deps, env, msg, Amount::Native(coin), info.sender)
+        }
+    }
+}
+
+pub fn execute_andr_receive(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: AndromedaMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        AndromedaMsg::Receive(data) => {
+            let transfer_msg: TransferMsg = parse_message(&data)?;
+            let coin = match info.funds.len() {
+                0 => Err(PaymentError::NoFunds {}),
+                1 => {
+                    let coin = &info.funds[0];
+                    if coin.amount.is_zero() {
+                        Err(PaymentError::NoFunds {})
+                    } else {
+                        Ok(coin.clone())
+                    }
+                }
+                _ => Err(PaymentError::MultipleDenoms {}),
```

Comment:
Since this is also used in the above `execute`, we should extract it into a helper function.